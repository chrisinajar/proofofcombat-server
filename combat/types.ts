import {
  CombatEntry,
  AttackType,
  HeroStats,
  InventoryItemType,
  EnchantmentType,
  HeroClasses,
  HeroSkills,
  ArtifactItem,
  InventoryItemImbue,
} from "types/graphql";

import type { Unit } from "../calculations/units/unit";

export type Attribute = keyof HeroStats;

export type Combatant = {
  level: number;
  name: string;
  attackType: AttackType;
  class: HeroClasses;
  equipment: CombatantGear;
  attributes: HeroStats;
  skills?: HeroSkills;
  damageReduction: number;
  health: number;
  maxHealth: number;
  luck: {
    smallModifier: number;
    largeModifier: number;
    ultraModifier: number;
  };
  unit: Unit;

  // attack speed
  attackSpeed: number;
  attackSpeedRemainder: number;
};

export type EnchantedCombatant = Combatant & {
  percentageDamageIncrease: number;
  percentageDamageReduction: number;
  percentageEnchantmentDamageReduction: number;
  enchanted: true;
  bonusAccuracy: number;
  bonusDodge: number;
  bonusWeaponTiers: number;
  bonusShieldTiers: number;
  bonusArmorTiers: number;
  lifesteal: number;
  // inverse for MATHS
  mesmerizeChance: number;
  focusChance: number;
};

export type AttackAttributes = {
  toHit: Attribute;
  damage: Attribute;
  dodge: Attribute;
  damageReduction: Attribute;
};

export type CombatResult = {
  attackerCombatant: Combatant;
  victimCombatant: Combatant;
  victimDamage: number;
  attackerDamage: number;
  victimEnchantmentDamage: number;
  attackerEnchantmentDamage: number;
  victimHeal: number;
  attackerHeal: number;

  victimDied: boolean;
  attackerDied: boolean;
  victimIsMesmerized: boolean;
  attackerIsMesmerized: boolean;
  victimIsDead: boolean;
  attackerIsDead: boolean;
  log: CombatEntry[];
};

export type CombatGear = {
  type?: InventoryItemType;
  level: number;
  baseItem?: string;
  enchantment?: EnchantmentType | null;
  imbue?: InventoryItemImbue | null;
};

export type QuestItem = {
  name: string;
  baseItem: string;
};

export type CombatantGear = {
  armor: CombatGear[];
  weapons: CombatGear[];
  quests: QuestItem[];
  artifact?: ArtifactItem;
};
