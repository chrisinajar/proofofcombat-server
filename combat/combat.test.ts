import { AttackType, Hero } from "types/graphql";
import { didHit, createHeroCombatant } from "./";
import Databases from "../db";

function generateHero(): Hero {
  const hero = Databases.hero.upgrade({
    name: "test",
    password: "sadfasdf",
    location: {
      x: 0,
      y: 0,
      map: "default",
    },
  });
  hero.combat.health = hero.combat.maxHealth;

  return hero;
}

function getHitOdds(heroA, attackType, heroB) {
  let didHitCount = 0;
  for (let i = 0; i < 100000; ++i) {
    if (didHit(heroA, attackType, heroB)) {
      didHitCount += 1;
    }
  }

  return Math.round(didHitCount / 1000) / 100;
}

describe("combat", () => {
  [
    { stat: "strength", attackType: AttackType.Melee },
    { stat: "dexterity", attackType: AttackType.Ranged },
    { stat: "constitution", attackType: AttackType.Blood },
    { stat: "intelligence", attackType: AttackType.Wizard },
    { stat: "wisdom", attackType: AttackType.Elemental },
    { stat: "willpower", attackType: AttackType.Smite },
  ].forEach((entry) => {
    it(`increasing ${entry.stat} makes it easier to hit with ${entry.attackType}`, () => {
      const hero = generateHero();
      const heroCombatant = createHeroCombatant(hero);
      const hero2 = generateHero();
      const hero2Combatant = createHeroCombatant(hero2);
      const oddsBefore = getHitOdds(
        heroCombatant,
        entry.attackType,
        hero2Combatant
      );

      heroCombatant.attributes[entry.stat] *= 1.1;

      const oddsAfter = getHitOdds(
        heroCombatant,
        entry.attackType,
        hero2Combatant
      );
      expect(oddsBefore).toBeLessThan(oddsAfter);
    });
  });
});
