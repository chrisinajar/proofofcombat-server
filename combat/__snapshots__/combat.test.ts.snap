// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds archer great items can destroy 28's 1`] = `
{
  "constitution": 5005,
  "dexterity": 262121,
  "intelligence": 5005,
  "luck": 12005,
  "strength": 5005,
  "willpower": 7764,
  "wisdom": 5005,
}
`;

exports[`builds archer normal items can kill level 5 mobs 1`] = `
{
  "constitution": 10,
  "dexterity": 152,
  "intelligence": 10,
  "luck": 17,
  "strength": 10,
  "willpower": 11,
  "wisdom": 10,
}
`;

exports[`builds archer trash items can kill level 1 mobs 1`] = `
{
  "constitution": 5,
  "dexterity": 5,
  "intelligence": 5,
  "luck": 5,
  "strength": 5,
  "willpower": 5,
  "wisdom": 5,
}
`;

exports[`builds archer uber items can deal destroy 32s at max level 1`] = `
{
  "constitution": 8008,
  "dexterity": 293893,
  "intelligence": 8008,
  "luck": 19208,
  "strength": 8008,
  "willpower": 24846,
  "wisdom": 16016,
}
`;

exports[`builds melee great items can destroy 28's 1`] = `
{
  "constitution": 7007,
  "dexterity": 87554,
  "intelligence": 7007,
  "luck": 8505,
  "strength": 169438,
  "willpower": 13044,
  "wisdom": 7007,
}
`;

exports[`builds melee normal items can kill level 5 mobs 1`] = `
{
  "constitution": 10,
  "dexterity": 60,
  "intelligence": 10,
  "luck": 14,
  "strength": 108,
  "willpower": 13,
  "wisdom": 10,
}
`;

exports[`builds melee trash items can kill level 1 mobs 1`] = `
{
  "constitution": 5,
  "dexterity": 7,
  "intelligence": 5,
  "luck": 5,
  "strength": 8,
  "willpower": 6,
  "wisdom": 5,
}
`;

exports[`builds melee uber items can deal destroy 32s at max level 1`] = `
{
  "constitution": 11512,
  "dexterity": 125077,
  "intelligence": 11512,
  "luck": 16160,
  "strength": 350978,
  "willpower": 21430,
  "wisdom": 11512,
}
`;

exports[`calculateEnchantmentDamage should calculate damage correctly 1`] = `
{
  "attackerDamage": 0,
  "attackerHeal": 473749,
  "victimDamage": 0,
  "victimHeal": 473749,
}
`;

exports[`calculateEnchantmentDamage should calculate damage correctly 2`] = `
{
  "attackerDamage": 0,
  "attackerHeal": 474130,
  "victimDamage": 192,
  "victimHeal": 473749,
}
`;

exports[`calculateEnchantmentDamage should calculate damage correctly 3`] = `
{
  "attackerDamage": 0,
  "attackerHeal": 474712,
  "victimDamage": 585,
  "victimHeal": 473749,
}
`;

exports[`calculateEnchantmentDamage should calculate damage correctly 4`] = `
{
  "attackerDamage": 0,
  "attackerHeal": 476119,
  "victimDamage": 1389,
  "victimHeal": 473749,
}
`;
